### PR TITLE
revive idba_ud

### PR DESCRIPTION
**Pull request by bernt-matthias**
One of our users needs idba_ud .. so I use the chance to revive the galaxy tool. The commit is based on https://toolshed.g2.bx.psu.edu/repos/stheil/idba_ud_wrapper (Revision: 2:edda44d54b59). Currently I only switched to conda, added test, help, and citation. 

Discussion: 
- our users needs a patched version of ibda_ud (she needed to change one of the hard coded limits .. I remember that it was the length of 128 that separated short and long reads). Do you see a way to realize this in conda/Galaxy
- do we want/need to keep the fq2fa tool (for interleaving and N removal)?

### CODE REVIEW

File: tools/idba_ud/fq2fa.xml [hunk 1]
```diff
@@ -0,0 +1,55 @@
+<tool id="fq2fa" name="fq2fa" version="1.1.1 (Oct 18, 2012)">
```
Comment (bgruening) [inline tools/idba_ud/fq2fa.xml]: please remove the data from the version

File: tools/idba_ud/fq2fa.xml [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+  <description>fq2fa is a simple tool given by IDBA package that allows to convert fastq to fasta, merge paired-end reads and filter out reads containing 'N's</description>
+  <command>
```
Comment (bgruening) [inline tools/idba_ud/fq2fa.xml]: please add a requirement

File: tools/idba_ud/fq2fa.xml [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+    fq2fa
+    #if $fileLayout.layout=='paired'
+      --paired $fileLayout.sequence
```
Comment (bgruening) [inline tools/idba_ud/fq2fa.xml]: data params and text params should be single quoted

File: tools/idba_ud/fq2fa.xml [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+    #else if $fileLayout.layout=='merge'
+      --merge $fileLayout.sequenceR1 $fileLayout.sequenceR2
+    #else
+      $fileLayout.sequence
+    #end if
+    $filter
+    $output
```
Comment (bgruening) [inline tools/idba_ud/fq2fa.xml]: single quotes

File: tools/idba_ud/fq2fa.xml [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+        <option value="merge">paired-end in two files</option>
+      </param>
+      <when value="">
+        <param name="sequence" type="data" format="fastq" label="Single sequence file"/>
+      </when>
+      <when value="--paired">
+        <param name="sequence" type="data" format="fastq" label="Interleaved sequence file"/>
+      </when>
+      <when value="--merge">
+        <param name="sequenceR1" type="data" format="fastq" label="Forward sequence file (R1)"/>
```
Comment (bgruening) [inline tools/idba_ud/fq2fa.xml]: only fastq or also fastq.gz?

File: tools/idba_ud/fq2fa.xml [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+        <param name="sequenceR2" type="data" format="fastq" label="Reverse sequence file (R2)"/>
+      </when>
+    </conditional>
+    <param name="filter" type="boolean" checked="false" truevalue="--filter" falsevalue="" label="Filter out reads containing 'N'"/>
+  </inputs>
+  <outputs>
+		<data name="output" format="fasta" label="Fasta from fq2fa"/>
```
Comment (bgruening) [inline tools/idba_ud/fq2fa.xml]: please adjust the indentation ... 4 spaces like python

File: tools/idba_ud/idba_ud.xml [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+	--similar $similar
+	--max_mismatch $max_mismatch
+	--min_pairs $min_pairs
+	#if str( $other ) != "None":
+		$other
+	#end if
+
+	; 
+
+	mv out/scaffold.fa $output
```
Comment (bgruening) [inline tools/idba_ud/idba_ud.xml]: quotes

File: tools/idba_ud/idba_ud.xml [hunk 8]
```diff
@@ -0,0 +1,99 @@
...
+
+		<param name="mink" type="integer" value="20" max="124" label="Minimum k value. Lower or equal to 124 (--mink)"/>
+		<param name="maxk" type="integer" value="100" max="124" label="Maximum k value. Lower or equal to 124 (--maxk)"/>
+		<param name="step" type="integer" value="20" label="Increment of k-mer of each iteration (--step)"/>
+		<param name="inner_mink" type="integer" value="10" label="Inner minimum k value (--inner_mink)"/>
+		<param name="inner_step" type="integer" value="5" label="Inner increment of k-mer (--inner_step)"/>
+		<param name="prefix" type="integer" value="3" label="Prefix length used to build sub k-mer table (--prefix)"/>
+		<param name="min_count" type="integer" value="2" label="Minimum multiplicity for filtering k-mer when building the graph (--min_count)"/>
+		<param name="min_support" type="integer" value="1" label="Minimum supoort in each iteration (--min_support)"/>
+		<param name="num_threads" type="integer" value="0" label="Number of threads (--num_threads)"/>
```
Comment (bgruening) [inline tools/idba_ud/idba_ud.xml]: you could use arguments= here everywhere ...

**Comment by nsoranzo** [comment]
> do we want/need to keep the fq2fa tool (for interleaving and N removal)?

Not sure it is worth to have, since we have https://toolshed.g2.bx.psu.edu/view/devteam/fastq_to_fasta

**Comment by bernt-matthias** [comment]
But: can it also do the interleaving in the same way?

**Comment by nsoranzo** [comment]
No, for that I'd need to use a separate tool like  https://toolshed.g2.bx.psu.edu/view/devteam/fastq_paired_end_interlacer

File: tools/idba_ud/idba_ud.xml [hunk 9]
```diff
@@ -0,0 +1,99 @@
...
+    --prefix $prefix
+    --min_count $min_count
+    --min_support $min_support
+    --num_threads \${GALAXY_SLOTS:-1}
+    --seed_kmer $seed_kmer
+    --min_contig $min_contig
+    --similar $similar
+    --max_mismatch $max_mismatch
+    --min_pairs $min_pairs
+    #if str( $other ) != "None":
```
Comment (bgruening) [inline tools/idba_ud/idba_ud.xml]: I think you can just use `#if $other:`.

File: tools/idba_ud/idba_ud.xml [hunk 10]
```diff
@@ -0,0 +1,99 @@
...
+        $other
+    #end if
+
+    ;
+
+    mv out/scaffold.fa '$output'
```
Comment (bgruening) [inline tools/idba_ud/idba_ud.xml]: you can remove this and use `from_work_dir`

File: tools/idba_ud/idba_ud.xml [hunk 11]
```diff
@@ -0,0 +1,99 @@
...
+
+    </command>
+
+    <inputs>
+        <param name="read" type="data" format="fasta" argument="--read" label="Fasta read file. Lower or equal to 250b"/>
+        <param name="long_read" type="data" format="fasta" optional="true" argument="--long_read" label="Fasta long read file. More than 250b"/>
```
Comment (bgruening) [inline tools/idba_ud/idba_ud.xml]: if you use `arguement` in this way, you can remove name, if you like.

File: tools/idba_ud/idba_ud.xml (hunk 9 above)
Comment (bernt-matthias) [reply tools/idba_ud/idba_ud.xml]: done.

File: tools/idba_ud/idba_ud.xml (hunk 10 above)
Comment (bernt-matthias) [reply tools/idba_ud/idba_ud.xml]: good idea. thanks

File: tools/idba_ud/idba_ud.xml (hunk 11 above)
Comment (bernt-matthias) [reply tools/idba_ud/idba_ud.xml]: cool. done.

**Comment by bernt-matthias** [comment]
Also started to document how to generate interleaved fasta. Question remains: how to remove 'N's and do we want users to execute 2-3 steps? Wrt. testing these steps: one of our users who requested idba could test it right away.

**Comment by bgruening** [comment]
> And another question: can/should we add the galaxy tools to the requirements?
What? :)

**Comment by bernt-matthias** [comment]
OK. I removed fq2fa and updated the help (is there a nice way for rendering the links?). I did not mention the potential problem of non-standard bases .. hoping that they will be removed by quality filtering anyway.

File: tools/idba_ud/idba_ud.xml [hunk 12]
```diff
@@ -0,0 +1,107 @@
...
+i.e. paired-end reads need to be stored in the same FASTA file suc h that a pair 
+of reads should be in two consecutive lines. 
+In Galaxy paired reads in separate FASTQ files can be converted into interleaved 
+FASTA using the tools:  
+
+* *FASTQ interlacer on paired end read* (https://toolshed.g2.bx.psu.edu/view/devteam/fastq_paired_end_interlacer)
+* *Samtools extract FASTA or FASTQ from a SAM file* (https://toolshed.g2.bx.psu.edu/view/devteam/fastq_to_fasta)
+
+Note that, IDBA-UD assumes that the paired-end reads are in order (-&gt;, &lt;-). 
+If your data is in reverse order (&lt;-, -&gt;), please convert it by yourself.
```
Comment (nsoranzo) [inline tools/idba_ud/idba_ud.xml]: If you [CDATA](https://galaxy-iuc-standards.readthedocs.io/en/latest/best_practices/tool_xml.html#help-tag) you don't need to use HTML entities like `&gt;`

File: tools/idba_ud/idba_ud.xml [hunk 13]
```diff
@@ -0,0 +1,107 @@
+<tool id="idba_ud" name="IDBA-UD" version="1.1.3">
+    <description>
+        Iterative de Bruijn Graph Assembler <!--for sequencing data with highly uneven depth-->
+    </description>
+    <macros>
+        <import>macros.xml</import>
+    </macros>
+    <expand macro="requirements"/>
+    <command>
```
Comment (nsoranzo) [inline tools/idba_ud/idba_ud.xml]: CDATA

File: tools/idba_ud/idba_ud.xml [hunk 14]
```diff
@@ -0,0 +1,107 @@
...
+
+
+Input: IDBA-UD takes interleaved paired end data in the FASTA format as input, 
+i.e. paired-end reads need to be stored in the same FASTA file suc h that a pair 
+of reads should be in two consecutive lines. 
+In Galaxy paired reads in separate FASTQ files can be converted into interleaved 
+FASTA using the tools:  
+
+* *FASTQ interlacer on paired end read* (https://toolshed.g2.bx.psu.edu/view/devteam/fastq_paired_end_interlacer)
+* *Samtools extract FASTA or FASTQ from a SAM file* (https://toolshed.g2.bx.psu.edu/view/devteam/fastq_to_fasta)
```
Comment (nsoranzo) [inline tools/idba_ud/idba_ud.xml]: The help section is interpreted as [reStructuredText](http://docutils.sourceforge.net/rst.html), so you can use the syntax for [embedded URIs](http://docutils.sourceforge.net/docs/user/rst/quickref.html#hyperlink-targets).

File: tools/idba_ud/idba_ud.xml (hunk 12 above)
Comment (bernt-matthias) [reply tools/idba_ud/idba_ud.xml]: done

File: tools/idba_ud/idba_ud.xml (hunk 14 above)
Comment (bernt-matthias) [reply tools/idba_ud/idba_ud.xml]: done

**Comment by bernt-matthias** [comment]
I fixed the last todos. Will meet a user next week who may tell me if the tool integrates nicely with the others as documented in the help. If so then the tool is ready to merge from my side.

**Comment by bgruening** [comment]
@bernt-matthias can you ping us if you got the feedback or should we give this now a final review?

**Comment by bernt-matthias** [comment]
Users are happy. Review is always a good idea .. if you have the time for it .. maybe diff against the original version.

**Comment by bernt-matthias** [comment]
ready to merge from my side

**Comment by bgruening** [comment]
Danke Matthias!

**Pull request merged**